### PR TITLE
feat(transformer/typescript): remove type only imports/exports correctly

### PR DESCRIPTION
**Pull request by Dunqing**

### CODE REVIEW

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #2055** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2055?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #2056** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2056?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2055?utm_source=stack-comment).

File: crates/oxc_transformer/src/typescript/mod.rs [hunk 1]
```diff
@@ -109,24 +102,84 @@ impl<'a> TypeScript<'a> {
     /// * Adds `export {}` if all import / export statements are removed, this is used to tell
     /// downstream tools that this file is in ESM.
     pub fn transform_program(&mut self, program: &mut Program<'a>) {
-        let mut needs_explicit_esm = false;
+        let mut export_type_names = FxHashSet::default();
+        let mut export_names = FxHashSet::default();
 
-        for stmt in program.body.iter_mut() {
+        // Collect export names
+        program.body.iter().for_each(|stmt| {
```
Comment (Boshen) [inline crates/oxc_transformer/src/typescript/mod.rs:109]: We already have them inside `ModuleRecord`, I guess these are transformed by previous transformers?

Comment (Dunqing) [reply crates/oxc_transformer/src/typescript/mod.rs:109]: Yes, But I worry if other plugins change the imports/exports, the ModuleRecord is not updated. How do you think I could fix this? Like `nodes` / `scopes` / `symbols` do? So it seems that whenever ast changes, we need to update everything in `Semantic`

Comment (Boshen) [reply crates/oxc_transformer/src/typescript/mod.rs:109]: Let's get the tests passing first and refactor later

**Comment by Dunqing** [comment]
### Merge activity

* **Jan 17, 2:57 AM**: @Dunqing started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2055).

**Pull request merged**